crates/burn-ndarray/src/ops/matmul.rs

```diff
@@ -11,101 +12,270 @@ pub(crate) fn matmul<E, const D: usize>(
-            panic!("Broadcast on multiple dimensions is not yet supported");
-        }
-
-        if batch_size_rhs > batch_size && batch_size_rhs != 1 {
-            panic!("Broadcast on multiple dimensions is not yet supported");
-        }
-
-        let alpha: E = 1.0.elem();
-        let beta: E = 0.0.elem();
-
-        let mut out_array = ndarray::Array3::<E>::zeros((batch_size, m, n));
+    let out: NdArrayTensor<E, D> = run_par!(|| {
+        let mut out_array = ndarray::Array3::<E>::zeros((num_out_batches, m, n));
         let unsafe_shared_out_array = UnsafeSharedRef::new(&mut out_array);
 
-        let lhs_array = lhs.array.into_shape((batch_size_lhs, m, k)).unwrap();
-        let rhs_array = rhs.array.into_shape((batch_size_rhs, k, n)).unwrap();
+        let lhs_array = NdArray::<E>::float_reshape(lhs, Shape::new([num_l_batches, m, k])).array;
+        let rhs_array = NdArray::<E>::float_reshape(rhs, Shape::new([num_r_batches, k, n])).array;
+
+        iter_range_par!(0, num_out_batches).for_each(|out_batch| {
+            // Here, we:
+            //   1. Un-flatten the output batch into a component-based batch index.
+            //   2. Use the strides for left and right batch indices to convert it to a flattened
+            //      batch for left and right.
+            let out_index = strides_out.unflatten(out_batch);
+            let l_batch = strides_lhs.flatten(&out_index);
+            let r_batch = strides_rhs.flatten(&out_index);
 
-        iter_range_par!(0, batch_size).for_each(|b| {
-            let lhs_slice = match batch_size_lhs == 1 {
-                true => lhs_array.slice(s!(0, .., ..)),
-                false => lhs_array.slice(s!(b, .., ..)),
-            };
-            let rhs_slice = match batch_size_rhs == 1 {
-                true => rhs_array.slice(s!(0, .., ..)),
-                false => rhs_array.slice(s!(b, .., ..)),
-            };
+            let lhs_slice = lhs_array.slice(s!(l_batch, .., ..));
+            let rhs_slice = rhs_array.slice(s!(r_batch, .., ..));
 
             unsafe {
-                let mut out_slice = unsafe_shared_out_array.get().slice_mut(s!(b, .., ..));
+                let mut out_slice = unsafe_shared_out_array
+                    .get()
+                    .slice_mut(s!(out_batch, .., ..));
 
                 ndarray::linalg::general_mat_mul(
                     alpha,
                     &lhs_slice,
                     &rhs_slice,
                     beta,
                     &mut out_slice,
-                );
+                )
             }
         });
 
         NdArrayTensor::new(out_array.into_shared().into_dyn())
-    })
+    });
+
+    NdArray::<E>::float_reshape(out, out_shape)
 }
 
-fn reshape<E: FloatNdArrayElement, const D: usize>(
-    tensor: NdArrayTensor<E, D>,
-) -> NdArrayTensor<E, 3> {
-    let shape = tensor.shape();
+#[derive(Debug, PartialEq)]
+struct Strides {
+    strides: Vec<usize>,
+}
```

Comment:
The `Vec` here is a little frustrating. I gather that with nightly, it is possible to write `[usize; {D - 2}]` or similar. However, I've taken care to allocate these with a fixed capacity, so hopefully they're not too much slower than a sized array.